### PR TITLE
Add a reverse migration to distribution change

### PR DESCRIPTION
**Pull request by mdellweg**

### CODE REVIEW

**Comment by mdellweg** [comment]
It turns out, that this operation fails to perform backwards. I have no idea yet, how to solve this.

``` python
        migrations.AlterField(
            model_name='filedistribution',
            name='distribution_ptr',
            field=models.OneToOneField(auto_created=True,
                                       on_delete=django.db.models.deletion.CASCADE,
                                       parent_link=True, primary_key=True,
                                       related_name='file_filedistribution', serialize=False,
                                       to='core.Distribution'),
            preserve_default=False,
        ),
```

**Comment by mdellweg** [comment]
It turns out, that the following statements in the resulting sql-backward migration (pulpcore-manager sqlmigrate file 0009 --backwards) are in the wrong order:
``` sql
ALTER TABLE "file_filedistribution" ALTER COLUMN "distribution_ptr_id" DROP NOT NULL;
ALTER TABLE "file_filedistribution" DROP CONSTRAINT "file_filedistribution_distribution_ptr_id_5ce5564b_pk";
```
I think, this is a django bug.

**Comment by daviddavis** [comment]
This looks good to me.

**Comment by mdellweg** [comment]
> This looks good to me.

It does not work. Once loaded with data, you get:
django.db.utils.InternalError: could not find trigger 86175

**Comment by mdellweg** [comment]
This should work now. I tested forward and backward several times with the following data:

```
#!/bin/bash

set -eux

pulp file distribution destroy --name "test2" || true
pulp file distribution destroy --name "test1" || true
pulp file repository destroy --name "test1" || true
pulp file remote destroy --name "test1" || true

pulp file remote create --name "test1" --url "https://fixtures.pulpproject.org/file/PULP_MANIFEST"
pulp file repository create --name "test1" --remote "test1"
pulp file repository sync --name "test1"
PUBLICATION_HREF=$(pulp file publication create --repository "test1" | jq -r '.pulp_href')
pulp file distribution create --name "test1" --base-path "test1"
pulp file distribution create --name "test2" --base-path "test2" --publication $PUBLICATION_HREF
```

**Comment by daviddavis** [comment]
I created some distributions against `1c96911` then checked out this PR and ran the migration. Everything worked. So then I ran `pulpcore-manager migrate file 0008` followed by `pulpcore-manager migrate file 0009` and confirmed everything worked again. :+1:

**Review by daviddavis — APPROVED** [review]

**Comment by daviddavis** [comment]
I guess we need to update this @pulp/rpm migration too?

https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/migrations/0032_new_distribution_model.py

**Comment by dralley** [comment]
@daviddavis Yup

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py [hunk 1]
```diff
@@ -4,73 +4,74 @@
...
-pks_to_delete = []
-
-
-def migrate_data_from_old_master_model_to_new_master_model(apps, schema_editor):
+def migrate_data_from_old_model_to_new_model_up(apps, schema_editor):
+    """ Move objects from FileDistribution to NewFileDistribution."""
     FileDistribution = apps.get_model('file', 'FileDistribution')
-    CoreDistribution = apps.get_model('core', 'Distribution')
-    for old_file_distribution in FileDistribution.objects.all():
+    NewFileDistribution = apps.get_model('file', 'NewFileDistribution')
```
Comment (dralley) [inline pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:10]: Is this just an attempt at making it more generic and proper or is it strictly necessary for functional reasons for the file plugin?

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py [hunk 2]
```diff
@@ -4,73 +4,74 @@
...
-                name=old_file_distribution.name,
-                base_path=old_file_distribution.base_path,
-                content_guard=old_file_distribution.content_guard,
-                remote=old_file_distribution.remote,
-                publication=old_file_distribution.publication,
-            )
-            new_master_model_entry.save()
-            old_file_distribution.distribution_ptr = new_master_model_entry
-            old_file_distribution.save()
-            pks_to_delete.append(old_file_distribution.pulp_id)
```
Comment (dralley) [inline pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:29]: I'm glad we can get rid of this, I'm just concerned that perhaps there was some necessary reason it was done in such a convoluted fashion originally?

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py [hunk 3]
```diff
@@ -4,73 +4,74 @@
...
+                base_path=file_distribution.base_path,
+                content_guard=file_distribution.content_guard,
+                remote=file_distribution.remote,
+                publication=file_distribution.publication
+            ).save()
+            file_distribution.delete()
 
 
 class Migration(migrations.Migration):
+    atomic = False
```
Comment (dralley) [inline pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:48]: Why is this necessary?

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 1 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:10]: Not sure, what you mean. The `apps.get_model` is Django's way to provide you the historical model with a schema that matches the current state of the migration.
In this case, it provides you the transient `NewFileDistribution` that only pops into existance to later be renamed to `FileDistribution`. Its evanescent nature manifests in its complete lack of presence in the actual code.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 2 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:29]: I believe the problem of being unable to delete the old base objects was, that the `basedistribution_base_ptr` had a foreign key constraint that was only lifted by removing the field from the table.
In the new approach there is no such ambiguity of inheritance that prevents the immediate deletion of old objects.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 3 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:48]: This solves the aforementioned trigger error.
My theory: When inserting new items into the new table, Django or Postgres add a trigger to check for constraints upon leaving the transaction. Renaming the table however breaks the relation and the trigger cannot be found anymore. Ending the transaction earlier fixes the problem.
Django suggests making data migrations non atomic:
https://docs.djangoproject.com/en/2.2/howto/writing-migrations/#non-atomic-migrations

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 1 above)
Comment (dralley) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:10]: What I mean is that NewFileDistribution and core.Distribution are effectively the same synce the former adds no new fields over the latter.  I agree that it should be used since it's more specific, I'm just trying to make sure my mental model is correct.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 2 above)
Comment (dralley) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:29]: +1

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 3 above)
Comment (dralley) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:48]: I know that constraints are checked last, I don't know the specific implementation details.  But this sounds plausible. It also sounds potentially like a bug - maybe we should bring it up in #django and what people say.

Are there portions of the migration that we can make atomic or non-atomic selectively rather than making the entire migration non-atomic in a blanket way?

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py [hunk 4]
```diff
@@ -4,73 +4,74 @@
...
+            options={
+                'default_related_name': '%(app_label)s_%(model_name)s',
+            },
+            bases=('core.distribution',),
         ),
-        migrations.RunPython(migrate_data_from_old_master_model_to_new_master_model),
-        migrations.RemoveField(
-            model_name='filedistribution',
-            name='basedistribution_ptr',
+        migrations.RunPython(
```
Comment (dralley) [inline pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:66]: What I mean is, I believe we can put `atomic=True` here rather than having our own transactions inside, and perhaps for other operations also.

Sidenote: if this is declared atomic by default, it means that the "transaction" in the migration was actually just a "savepoint".  I don't remember how the semantic differ, precisely.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 1 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:10]: NewFileDistribution here is more than core.Distribution, as it's a detail table. So i am copying objects from one detail table to the other.
IMHO you should never add objects to the core.Distribution table directly.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 4 above)
Comment (ggainey) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:66]: transactions in postgres can have multiple savepoints, and rollback can go back to a specific savepoint (rather than rolling back the entire transaction). You have to specifically do a 'ROLLBACK TO savepoint', iirc, to take advantage of this, and I don't know if/how Django does so.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 3 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:48]: I have not seen a way to make several steps of the migration atomic.
What we could do is move the last two steps (Deleting and renaming Tables) in a separate migration. But i am unsure whether this would gain much.

File: pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py (hunk 4 above)
Comment (mdellweg) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:66]: I had atomic=true with and without the inner transactions and it produced the same error.

Comment (dralley) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:66]: Good to know

Comment (dralley) [reply pulp_file/app/migrations/0009_move_data_to_new_master_distribution_model.py:66]: I'm still interested in figuring out of this is an upstream bug, but LGTM

**Review by dralley — APPROVED** [review]

**Comment by dralley** [comment]
@mdellweg I'm planning on swapping the distribution models today on many of the plugins.  If you can copy this PR to the RPM plugin and fix the template to use the new model, then I'll handle all of the rest.

**Comment by mdellweg** [comment]
> @mdellweg I'm planning on swapping the distribution models today on many of the plugins. If you can copy this PR to the RPM plugin and fix the template to use the new model, then I'll handle all of the rest.

What template do you mean?

**Comment by dralley** [comment]
@mdellweg The plugin template. e.g. https://github.com/pulp/plugin_template/blob/master/templates/bootstrap/plugin_name/app/viewsets.py.j2#L218

**Comment by daviddavis** [comment]
Oh yea, I totally forgot about the plugin template. I can take care of it unless @mdellweg you want to.

**Comment by mdellweg** [comment]
@dralley https://github.com/pulp/pulp_rpm/pull/1971

@daviddavis that would be great!
